### PR TITLE
ath79: mikrotik: add support for early RB911 Lite5 NAND boards

### PR DESCRIPTION
**Pull request by rogerpueyo**
These two patches add support for the early version of the MikroTik RB911 Lite5 boards with NAND storage. They were very almost hardware-identical to the already supported SXT Lite5, even the bootloader passes the same board identifier "sxt5n". The only difference is found in the tiny SPI NOR storage for the bootloader, which is half the size (64 kB for RB911Lite5NAND, 128 kB for the SXTLite5).

The first patch adapts the "SXT 5N" .dtsi to handle both 64 kB and 128 kB (bootloader2 partition will be added on a per-device basis, when needed). The second patch adds actual support for the RB911Lite5NAND boards.

Later versions of the RB911Lite5, which used an SPI NOR chip as their main storage, are not supported by the "sxt5n" platform, but by the "911L" one. The whole discussion and findings can be found at https://bugs.openwrt.org/index.php?do=details&task_id=3471.

### CODE REVIEW

File: target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts [hunk 1]
```diff
@@ -6,3 +6,11 @@
 	compatible = "mikrotik,routerboard-sxt-5nd-r2", "qca,ar9344";
 	model = "MikroTik RouterBOARD SXT 5nD r2 (SXT Lite5)";
 };
+
+&routerboot {
+	bootloader2: partition@10000 {
```
Comment (adschm) [inline target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts:11]: Is the bootloader2: DT label needed for something I don't see, or can it be dropped?

**Comment by adschm** [comment]
Has this early device been sold regularly?

Note that the Signed-by in your second patch is duplicated.

File: target/linux/ath79/mikrotik/base-files/etc/board.d/01_leds [hunk 2]
```diff
@@ -10,6 +10,7 @@ case "$board" in
 mikrotik,routerboard-lhg-2nd)
 	ucidef_set_led_netdev "lan" "lan" "green:lan" "eth0"
 	;;
+mikrotik,routerboard-911-5hn-nand|\
```
Comment (adschm) [inline target/linux/ath79/mikrotik/base-files/etc/board.d/01_leds:13]: Block needs to be moved to keep sorting.

**Comment by adschm** [comment]
So, am I understanding this right:
The early version uses NAND and the later one doesn't. I.e., one could name the devices with "early" and "late", or with "nand" and "spi" as you seem to do it? So, early=nand and late=spi-only ...?

File: target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts (hunk 1 above)
Comment (rogerpueyo) [reply target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts:11]: Dropped, thanks.

File: target/linux/ath79/mikrotik/base-files/etc/board.d/01_leds (hunk 2 above)
Comment (rogerpueyo) [reply target/linux/ath79/mikrotik/base-files/etc/board.d/01_leds:13]: True. Thanks.

File: target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5n.dtsi [hunk 3]
```diff
@@ -99,9 +99,9 @@
 			#address-cells = <1>;
 			#size-cells = <1>;
 
-			partition@0 {
+			routerboot: partition@0 {
 				label = "RouterBoot";
-				reg = <0x0 0x20000>;
+				reg = <0x0 0x0>;
```
Comment (adschm) [inline target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5n.dtsi:104]: @f00b4r0 Is setting the size to 0x0 here harmless? The last partition for a 64k chip would be soft_config then which has no size as well ...

**Comment by adschm** [comment]
You amended the change to the wrong commit.

File: target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5n.dtsi (hunk 3 above)
Comment (f00b4r0) [reply target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5n.dtsi:104]: now, this is invalid. The top-level "Routerboot" partition is a "fixed-partitions" style partition and its start offset and size must be correctly set.

**Comment by rogerpueyo** [comment]
> Note that the Signed-by in your second patch is duplicated.

Removed. Thanks.

> Has this early device been sold regularly?

I don't actually own the device. From the discussion we had at https://bugs.openwrt.org/index.php?do=details&task_id=3471 I understand it was sold regularly and, at some point, MikroTik slightly changed the hardware (just the storage, it seems) but kept selling it under the same name. This website from 2014 states it has NAND storage:

File: target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts (hunk 1 above)
Comment (f00b4r0) [reply target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts]: Don't do this. Don't split partitions between dtsi and dts. This is going to be a maintenance nightmare.
Why do you need to do this to begin with? The 64kB devices do have a secondary bootloader.

**Review by f00b4r0 — CHANGES_REQUESTED** [review]
Replying to @adschm's ping.

Please note once again that supporting NAND devices is _BROKEN_ in current OpenWRT and will eventually lead to bricked devices.

File: target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts [hunk 4]
```diff
@@ -6,3 +6,11 @@
...
 	model = "MikroTik RouterBOARD SXT 5nD r2 (SXT Lite5)";
 };
+
+&routerboot {
+	partition@10000 {
+		label = "bootloader2";
+		reg = <0x10000 0x10000>;
+		read-only;
+	};
+};
```
Comment (f00b4r0) [inline target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts:16]: Please don't do this. This makes no sense, and will be a maintenance nightmare. If the partition tables differ between devices, each DTS should have the full table I think. @adschm opinion?

**Comment by f00b4r0** [comment]
> So, am I understanding this right:
> The early version uses NAND and the later one doesn't. I.e., one could name the devices with "early" and "late", or with "nand" and "spi" as you seem to do it? So, early=nand and late=spi-only ...?

We're back onto the Mikrotik naming nightmare, #3270 all over again. I'm afraid this will eventually demonstrate that 1 image per board doesn't work for this vendor 😛

File: target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts (hunk 4 above)
Comment (adschm) [reply target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts:16]: I don't have a problem with splitting in general. One can improve the situation by adding comments to the possibly "incomplete" setup in the DTSI. I'd be more cautious with the routerboot situation, though, since this harder to grasp than the simple list of partitions you typically have.

So, no, I don't reject this generally. However, with the answer on the other question with respect to the size, this option is dead anyway as the size requirement will also require to have two full definitions.

**Comment by rogerpueyo** [comment]
> Note that the Signed-by in your second patch is duplicated.

Removed. Thanks.

> You amended the change to the wrong commit.

True. Fixed, thanks.

> Has this early device been sold regularly?

This MikroTik brochure (~2014) states 128 MB NAND: https://www.cyberbajt.pl/download/routerboard/911/RB911_series-140422133903.pdf
Also this website from 2014 states it has NAND storage: https://www.roc-noc.com/mikrotik/routerboard/RB911-5Hn.html

Looking at the wayback machine, at some point between Nov 2015 https://web.archive.org/web/20151118192818/http://routerboard.com/RB911-5Hn and Dec 2015 they had already switched to SPI NOR.

I don't actually own the device. From the discussion we had at https://bugs.openwrt.org/index.php?do=details&task_id=3471 I understand it was sold regularly and, at some point, MikroTik slightly changed the hardware (just the storage, it seems) but kept selling it under the same name.

> no, this is invalid. The top-level "Routerboot" partition is a "fixed-partitions" style partition and its start offset and size must be correctly set.
>
> Please don't do this. This makes no sense, and will be a maintenance nightmare. If the partition tables differ between devices, each DTS should have the full table I think. @adschm opinion?

So, one partition table per device. :+1:

> So, am I understanding this right:
> The early version uses NAND and the later one doesn't. I.e., one could name the devices with "early" and "late", or with "nand"  and "spi" as you seem to do it? So, early=nand and late=spi-only ...?

Yes, "early" version has 64 kB SPI NOR + 128 MB NAND, "late" version has 16 MB SPI NOR. Both with the same model name. Glorious.

**Comment by f00b4r0** [comment]
> > The early version uses NAND and the later one doesn't. I.e., one could name the devices with "early" and "late", or with "nand"  and "spi" as you seem to do it? So, early=nand and late=spi-only ...?
> 
> Yes, "early" version has 64 kB SPI NOR + 128 MB NAND, "late" version has 16 MB SPI NOR. Both with the same model name. Glorious.

The problem with that naming scheme is that the average user will never know which applies to their device. So we'll probably need to maintain some table. Alternatively, I'd suggest e.g. "-nand128M" and "-nor16M" suffixes, they're quite more closely related to both the hardware and the way mikrotik presents their devices on their website. My 2c.

**Review by acoul — APPROVED** [review]

**Comment by acoul** [comment]
Greetings,

using the above patches on a recent ath79/trunk on my recently purchased rb911-nand device, provided a successful working image

Linux version 5.4.81 (alex@aifnis) (gcc version 9.3.0 (OpenWrt GCC 9.3.0 r15132-9c13513ce1)) #0 Sun Dec 6 08:23:21 2020

CPU0 revision is: 0001974c (MIPS 74Kc)
MIPS: machine is MikroTik RouterBOARD 911-5Hn NAND (911 Lite5)
SoC: Atheros AR9344 rev 2

jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
nand: device found, Manufacturer ID: 0x98, Chip ID: 0xf1
nand: Toshiba NAND 128MiB 3,3V 8-bit
nand: 128 MiB, SLC, erase size: 128 KiB, page size: 2048, OOB size: 64
3 fixed-partitions partitions found on MTD device ar934x-nand
Creating 3 MTD partitions on "ar934x-nand":
spi-nor spi0.0: mx25l512e (64 Kbytes)
1 fixed-partitions partitions found on MTD device spi0.0
Creating 1 MTD partitions on "spi0.0":

File: target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5n.dtsi (hunk 3 above)
Comment (tmn505) [reply target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5n.dtsi:104]: As @f00b4r0 said, fixed means no size auto-detection guaranteed ever, so it could break in the future. For reference the flash map:
````
dev:    size   erasesize  name
mtd0: 00040000 00020000 "booter"
mtd1: 003c0000 00020000 "kernel"
mtd2: 07c00000 00020000 "ubi"
mtd3: 00010000 00001000 "RouterBoot"
mtd4: 0000c000 00001000 "bootloader1"
mtd5: 00001000 00001000 "hard_config"
mtd6: 00001000 00001000 "bios"
mtd7: 00001000 00001000 "soft_config"
````

File: target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts (hunk 4 above)
Comment (tmn505) [reply target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts:16]: Inevitably the NOR variant will be added, which will further clobber the visibility if kept like this. For readability full flash table would be preferable.

File: target/linux/ath79/image/mikrotik.mk [hunk 5]
```diff
@@ -9,6 +9,15 @@ define Device/mikrotik_routerboard-493g
 endef
 TARGET_DEVICES += mikrotik_routerboard-493g
 
+define Device/mikrotik_routerboard-911-5hn-nand
+  $(Device/mikrotik_nand)
+  SOC := ar9344
+  DEVICE_MODEL := RouterBOARD 911-5Hn NAND (911 Lite5)
+  DEVICE_PACKAGES += rssileds kmod-gpio-beeper
```
Comment (tmn505) [inline target/linux/ath79/image/mikrotik.mk:16]: There is no beeper on 911 board on both variants. Please remove package and move the dtsi node to appropriate dts.

**Review by tmn505 — CHANGES_REQUESTED** [review]
Additionally there is ethernet LED which works on ar71xx image but not with ath79.

I wonder if image name shouldn't be changed to routerboard-911-xhn-nand, since 2Hn version will work with the same image.

Thanks.

**Comment by ynezz** [comment]
Ping.

**Comment by acoul** [comment]
if anyone is willing to provide a suggested patch, I will be happy to test it & report back

**Comment by f00b4r0** [comment]
Besides the naming issue I have no further comment. I am however of the opinion that we shouldn't add more Mikrotik NAND boards given the broken-by-design support implementation.

**Comment by acoul** [comment]
this board works wonderfully under (patched) ath79/trunk. I agree that MikroTik sucks in shamefully violating GPL again & again, but their hw combined with open source is quite worthy 

> root@OpenWrt:~# cat /proc/cpuinfo 
system type		: Atheros AR9344 rev 2
machine			: MikroTik RouterBOARD 911-5Hn NAND (911 Lite5)
processor		: 0
cpu model		: MIPS 74Kc V4.12
BogoMIPS		: 299.26
wait instruction	: yes
microsecond timers	: yes
tlb_entries		: 32
extra interrupt vector	: yes
hardware watchpoint	: yes, count: 4, address/irw mask: [0x0ffc, 0x0ffc, 0x0ffb, 0x0ffb]
isa			: mips1 mips2 mips32r1 mips32r2
ASEs implemented	: mips16 dsp dsp2
Options implemented	: tlb 4kex 4k_cache prefetch mcheck ejtag llsc dc_aliases perf_cntr_intr_bit perf
shadow register sets	: 1
kscratch registers	: 0
package			: 0
core			: 0
VCED exceptions		: not available
VCEI exceptions		: not available

> root@OpenWrt:~# cat /proc/interrupts 
           CPU0       
  4:       1305      MIPS   4  19000000.eth
  7:     793290      MIPS   7  timer
  9:         20      MISC   3  ttyS0
 12:     130675      INTC   0  ath9k
 21:      20886      MISC  21  ar934x-nand
ERR:         24

> root@OpenWrt:~# df
Filesystem           1K-blocks      Used Available Use% Mounted on
/dev/root                 2304      2304         0 100% /rom
tmpfs                    29856        60     29796   0% /tmp
/dev/ubi0_2             108788       932    103020   1% /overlay
overlayfs:/overlay      108788       932    103020   1% /
tmpfs                      512         0       512   0% /dev

File: target/linux/ath79/image/mikrotik.mk (hunk 5 above)
Comment (rogerpueyo) [reply target/linux/ath79/image/mikrotik.mk:16]: @adschm, @tmn505,

Thanks for the review.

> Alternatively, I'd suggest e.g. "-nand128M" and "-nor16M" suffixes,

> I wonder if image name shouldn't be changed to routerboard-911-xhn-nand, since 2Hn version will work with the same image.

I added the -nand128m suffix, and changed the name to mention the flash size. I also added the 2.4 GHz version 2Hn in `mikrotik.mk`:
```
define Device/mikrotik_routerboard-911-5hn-nand128m
  $(Device/mikrotik_nand)
  SOC := ar9344
  DEVICE_MODEL := RouterBOARD 911-5Hn NAND 128MB (911 Lite5)
  DEVICE_ALT0_VENDOR := MikroTik
  DEVICE_ALT0_MODEL := RouterBOARD 911-2Hn NAND 128MB (911 Lite2)
  DEVICE_PACKAGES += rssileds
  SUPPORTED_DEVICES += rb-911-5hn-nand
endef
TARGET_DEVICES += mikrotik_routerboard-911-5hn-nand128m
```

> There is no beeper on 911 board on both variants. Please remove package and move the dtsi node to appropriate dts.

I removed the beeper and moved the node to the SXT Lite 5 dts.

> Inevitably the NOR variant will be added, which will further clobber the visibility if kept like this. For readability full flash table would be preferable.
> As @f00b4r0 said, fixed means no size auto-detection guaranteed ever, so it could break in the future. For reference the flash map:

I created a whole SPI NOR partition table on each device.

> Additionally there is ethernet LED which works on ar71xx image but not with ath79.

Unfortunately, I can't verify this, since I don't own the device.

File: target/linux/ath79/mikrotik/base-files/etc/board.d/01_leds [hunk 6]
```diff
@@ -23,6 +24,9 @@ mikrotik,routerboard-sxt-5nd-r2)
 	ucidef_set_led_rssi "rssimediumhigh" "rssimediumhigh" "green:rssimediumhigh" "wlan0" "60" "100"
 	ucidef_set_led_rssi "rssihigh" "rssihigh" "green:rssihigh" "wlan0" "80" "100"
 	;;
+mikrotik,routerboard-lhg-2nd)
```
Comment (adschm) [inline target/linux/ath79/mikrotik/base-files/etc/board.d/01_leds:27]: Unrelated change?

**Comment by adschm** [comment]
If I understand correctly, the 0x0 size issue is not resolved yet either.

File: target/linux/ath79/mikrotik/base-files/etc/board.d/01_leds (hunk 6 above)
Comment (rogerpueyo) [reply target/linux/ath79/mikrotik/base-files/etc/board.d/01_leds:27]: Weird. Maybe I screwed it during a rebase, the board is already configured in line ~9. Removing it.

File: target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts (hunk 4 above)
Comment (rogerpueyo) [reply target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts]: OK, this was fixed, as each device got its partition table.

File: target/linux/ath79/image/mikrotik.mk (hunk 5 above)
Comment (rogerpueyo) [reply target/linux/ath79/image/mikrotik.mk]: Except for the Ethernet LED, which I can't test, I'd say everything else here is solved.

**Comment by rogerpueyo** [comment]
> If I understand correctly, the 0x0 size issue is not resolved yet either.

Just fixed it. 0x0 is now 0x20000 for the SXT5 (128 kB), 0x10000 for the 911Lite5.

Rebased.

File: target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5n.dtsi (hunk 3 above)
Comment (rogerpueyo) [reply target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5n.dtsi]: Actual size for each device is set now.

**Review by tmn505 — APPROVED** [review]
Tested the PR on both kernels and it's working fine. The eth LED is not mandatory, I also don't know yet how to enable it. The device name could be tuned as this image will also work for 2Hn and aside from 128 MiB NAND variant there's no other, so nand128 could be shortened to nand. I'll leave the decision of a name for the device to someone who'll be merging this.

**Comment by rogerpueyo** [comment]
> The device name could be tuned as this image will also work for 2Hn and aside from 128 MiB NAND variant there's no other, so nand128 could be shortened to nand. I'll leave the decision of a name for the device to someone who'll be merging this.

I don't have a strong opinion on this neither, both options are fine for me. :+1: 

Rebased current master which conflicted with the patch.
